File: 238-product-of-array-except-self/238-product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int right[] = new int[nums.length];
        right[right.length-1] = 1;
        
        for(int i=nums.length-2; i>=0; i--){
            right[i] = right[i+1] * nums[i+1];
        }
        
        int left = 1;
        
        for(int i=0; i<right.length; i++){
            right[i] = left * right[i];
            left = left * nums[i];
        }
        
        return right;
    }
}

File: 238-product-of-array-except-self/238-product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        right = [1] * len(nums)
        
        for i in reversed(range(len(nums) - 1)):
            right[i] = right[i+1] * nums[i+1]
        
        left = 1
        
        for i in range(len(nums)):
            right[i] = left * right[i]
            left = left * nums[i]
            
        return right